Async methods in ConnectionPool

zhenlineo opened the pull request:
Removed ambiguous `IConnection.Dispose` method as it means differently for different `IConnection` implementation:

* `SocketConneciton` - Close network connection directly.
* PooledConnection - Reset connection and Release connection back to pool. Requires network send and receive.
* `TransactionConnection` - Release `Delegated` connection
* `ClusterConnection` - Perform `Delegated.Dispose()`

Instead `IConnection.Close` and `IConnection.CloseAsync()` are added as replacement to `IConnection.Dispose`
Renamed old`IConnection.Close` to `IConnection.Destroy` to better describe the only action in this method is to close/destroy resources only directly in this method.
Removed some unnecessary async method in `SocketClient` and `TcpClient`

Commit:
Async methods in ConnectionPool

Removed ambiguous `IConnection.Dispose` method as it means differently for different `IConnection` implementation:

* `SocketConneciton` - Close network connection directly.
* PooledConnection - Reset connection and Release connection back to pool. Requires network send and receive.
* `TransactionConnection` - Release `Delegated` connection
* `ClusterConnection` - Perform `Delegated.Dispose()`

Instead `IConnection.Close` and `IConnection.CloseAsync()` are added as replacement to `IConnection.Dispose`
Renamed old`IConnection.Close` to `IConnection.Destroy` to better describe the only action in this method is to close/destroy resources only directly in this method.
Removed some unnecessary async method in `SocketClient` and `TcpClient`

zhenlineo commented on Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs:
```diff
@@ -225,6 +216,12 @@ protected virtual void Dispose(bool isDisposing)
+        public Task CloseAsync()
+        {
+            Close();
+            return Task.CompletedTask; // TODO verify this is the correct way to do it
```
@ali-ince What is the best way to impl a async method that actually do not need to wait for anything?
As you could see, this CloseAsync() on SocketConnection is the same as Close()

ali-ince replied:
@zhenlineo I thought to remove those Async methods on our second pass - as we really do not need them at all. Instead we should use Close().

zhenlineo commented on Neo4j.Driver/Neo4j.Driver/Internal/Transaction.cs:
```diff
@@ -223,13 +223,19 @@ public TransactionConnection(Transaction transaction, IConnection connection)
+            public override Task CloseAsync()
+            {
+                Close();
+                return Task.CompletedTask;
```
Same here

Commit:
Making session.RunAsync all async

ali-ince committed:
Fixes and new internal async methods

1. Implemented IOutputStream#WriteAsync & IWriter.WriteAsync methods,
2. Implemented Packer#PackAsync methods,
3. Implemented IMessageRequestHandler#....Async methods,
4. Implemented IRequestMessage#DispatchAsync method,
5. Added missing ConfigureAwait calls,
6. Replaced ContinueWith with await calls.
